mm2src/coins/lp_coins.rs

```diff
@@ -1410,25 +1486,178 @@ pub trait CoinAssocTypes {
+    pub maker_pub: &'a Coin::Pubkey,
+    /// Unique data of specific swap
+    pub swap_unique_data: &'a [u8],
+}
+
+pub struct RefundMakerPaymentArgs<'a, Coin: CoinAssocTypes + ?Sized> {
+    /// Maker payment tx
+    pub maker_payment_tx: &'a Coin::Tx,
+    /// Maker will be able to refund the payment after this timestamp
+    pub time_lock: u64,
+    /// The hash of the secret generated by taker, this us used for immediate refund
+    pub taker_secret_hash: &'a [u8],
+    /// The hash of the secret generated by maker, taker needs it to spend the payment
+    pub maker_secret_hash: &'a [u8],
+    /// Taker's secret
+    pub taker_secret: &'a [u8],
+    /// Taker's HTLC pubkey
+    pub taker_pub: &'a Coin::Pubkey,
+    /// Unique data of specific swap
+    pub swap_unique_data: &'a [u8],
+}
+
+pub struct SpendMakerPaymentArgs<'a, Coin: CoinAssocTypes + ?Sized> {
+    /// Maker payment tx
+    pub maker_payment_tx: &'a Coin::Tx,
+    /// Maker will be able to refund the payment after this timestamp
+    pub time_lock: u64,
+    /// The hash of the secret generated by taker, this us used for immediate refund
+    pub taker_secret_hash: &'a [u8],
+    /// The hash of the secret generated by maker, taker needs it to spend the payment
+    pub maker_secret_hash: &'a [u8],
+    /// The secret generated by maker, revealed when maker spends taker's payment
+    pub maker_secret: &'a [u8],
+    /// Maker's HTLC pubkey
+    pub maker_pub: &'a Coin::Pubkey,
+    /// Unique data of specific swap
+    pub swap_unique_data: &'a [u8],
+}
+
+/// Operations specific to maker coin in [Trading Protocol Upgrade implementation](https://github.com/KomodoPlatform/komodo-defi-framework/issues/1895)
+#[async_trait]
+pub trait MakerCoinSwapOpsV2: CoinAssocTypes + Send + Sync + 'static {
+    /// Generate and broadcast maker payment transaction
+    async fn send_maker_payment_v2(&self, args: SendMakerPaymentArgs<'_, Self>) -> Result<Self::Tx, TransactionErr>;
+
+    /// Validate maker payment transaction
+    async fn validate_maker_payment_v2(&self, args: ValidateMakerPaymentArgs<'_, Self>) -> ValidatePaymentResult<()>;
+
+    /// Refund maker payment transaction using timelock path
+    async fn refund_maker_payment_v2_timelock(&self, args: RefundPaymentArgs<'_>) -> Result<Self::Tx, TransactionErr>;
+
+    /// Refund maker payment transaction using immediate refund path
+    async fn refund_maker_payment_v2_secret(
+        &self,
+        args: RefundMakerPaymentArgs<'_, Self>,
+    ) -> Result<Self::Tx, TransactionErr>;
+
+    /// Spend maker payment transaction
+    async fn spend_maker_payment_v2(&self, args: SpendMakerPaymentArgs<'_, Self>) -> Result<Self::Tx, TransactionErr>;
+}
+
+#[derive(Display)]
+pub enum WaitForTakerPaymentSpendError {
+    #[display(
+        fmt = "Timed out waiting for taker payment spend, wait_until {}, now {}",
+        wait_until,
+        now
+    )]
+    Timeout {
+        wait_until: u64,
+        now: u64,
+    },
```

Comment:
I suggest using plain text date formats for errors as most people will likely do the timestamp conversion manually (as it's not quite human-readable).

Reply:
Could you please elaborate on the approach you would like to see here? As I can see, timestamp formatting is not a part of standard library, so we have to rely on crates like `chrono` for this, likely changing the `u64` to something else.

Reply:
Yes, std doesn't have timestamp parsing abstractions. If you don't want to use chrono, you can do this in the sqlite as well (by using SELECT DATETIME you can do various timestamp operations). But as far as I know we already have bunch of time crates already in our dependency tree, so using one of them should be fine.

> likely changing the u64 to something else.

Yeah, you can create a custom data-type something like "Datetime" (which contains String value) and with "From" trait you can implement that conversion.

Reply:
Chrono is in our deps tree already, I will use a type provided by it.

Reply:
Using `chrono::NaiveDateTime` in `WaitForTakerPaymentSpendError` turned out to a bulky handling:
- `NaiveDateTime::from_timestamp` is deprecated.
- `NaiveDateTime::from_timestamp_opt` expects `i64` argument and returns an `Option`.

So, I have to write the following code:
```rust
impl From<WaitForOutputSpendErr> for WaitForTakerPaymentSpendError {
    fn from(err: WaitForOutputSpendErr) -> Self {
        match err {
            WaitForOutputSpendErr::Timeout { wait_until, now } => WaitForTakerPaymentSpendError::Timeout {
                wait_until: NaiveDateTime::from_timestamp_opt(wait_until as i64, 0).expect("valid timestamp"),
                now: NaiveDateTime::from_timestamp_opt(now as i64, 0).expect("valid timestamp"),
            },
            WaitForOutputSpendErr::NoOutputWithIndex(index) => {
                WaitForTakerPaymentSpendError::InvalidInputTx(format!("Tx doesn't have output with index {}", index))
            },
        }
    }
}
```

It's not possible to gracefully handle additional error cases in `impl From`, so I have to use `as` and `expect/unwrap`. A wider-scope refactoring is needed, which might be too much to accomplish the goal. So, I prefer to keep the code as is.

Reply:
Created an issue for that https://github.com/KomodoPlatform/komodo-defi-framework/issues/2055